chat.sympl

```diff
@@ -183,177 +228,260 @@ def _guard_input(input_description: str, input_: str) -> None:
-        not_a_member : str = cvm.tx.key_alias
-        cvm.error(f'Member {not_a_member} does not belong to the room. Operation denied.')
-
-    if not std.contains_using(members, cvm.tx.key_alias, _str_eq):
-            not_a_member : str = cvm.tx.key_alias
-            cvm.error(f'Member {not_a_member} does not belong to the room. Operation denied.')
-
+    #perform invite checks
+    invite_to_room_checks(room, new_member)
+    
+    #modify contract storage
     room.members = room.members + [new_member]
     cvm.storage.put(Identifier('room'), room)
+
+    #creates an invite to room event
     invite_to_room_event = InviteToRoomEvent(room=room, inviter=cvm.tx.key_alias, invitee=new_member)
     cvm.create_event('InviteToRoomEvent', std.json(invite_to_room_event))
-    cvm.job_complete(std.json(invite_to_room_event))
+    return invite_to_room_event
+
+#runs checks to see if the caller has the correct permissions to invite the 
+#a new member to a room
+@helper
+def invite_to_room_checks(room: Room, new_member: KeyAlias) -> None:
+    caller : KeyAlias = cvm.tx.key_alias
+
+    #check if the caller is an owner
+    if not std.contains_using(room.owners, caller, _str_eq):
+        cvm.error(f"{caller} is not an owner of the room {room.channel}.")
+
+    #check if the invitee is already a member
+    if std.contains_using(room.members, new_member, _str_eq):
+        cvm.error(f"Member {new_member} already in room {room.channel}.")
+
+    #check if the room is deleted
+    if room.is_deleted:
+        cvm.error(f"{room.channel} is deleted.")
 
-@public_clientside
+@clientside_helper
 def _remove_from_room(room_channel: ChannelName, member_to_remove: KeyAlias) -> None:
     room = _get_room(room_channel)
-    mrs : str = member_to_remove
-    if mrs == cvm.tx.key_alias:
-        cvm.error(f"Cannot remove self from room.")
-    members = room.members
-    if not std.contains_using(members, member_to_remove, _str_eq):
-        member_to_remove_str : str = member_to_remove
-        room_channel_str : str = room_channel
-        cvm.error(f'Member {member_to_remove_str} not in room {room_channel_str}.')
+    #run checks on removing
+    remove_from_room_checks(room, member_to_remove, cvm.tx.key_alias)
+
+    #if the member to remove is an owner, we want to remove ownership
+    #privileges, and then promote a new owner if there are no
+    #owners left
+    if std.contains_using(room.owners, member_to_remove, _str_eq):
+       _demote_owner(room_channel,member_to_remove)
+
     with PostTxArgs(room_channel):
         _remove_from_room_execute(room_channel, member_to_remove)
-    cvm.remove_owner(room_channel, member_to_remove)
+
+    #rotate the keys so that all members of the chatroom will be able to 
+    #read future messages, and the member we are removing will not be able to 
+    #read future messages
     cvm.rotate_key(room_channel)
-    updated_members = room.members
-    for member in updated_members:
+    for member in room.members:
```

Comment:
No, because the `owners` list is always going to be a subset of members (i.e. owners are also stored in the members list), maybe this should be done differently?

Reply:
👍 got it! yeah, I missed that the flow is invite then promote! makes sense!